Feature/component def reconciliation

merryman opened the pull request:
Once completed, this branch will allow users to define components without code by pure direct manipulation.

TODO
- [x] implement reset button to undo changes via DM in a component
- [x] switch from blacklisting to whitelisting of changes that feed into reconciliation (only via halo + side bar + (inspector?)).
- [x] support interactive definition of components
- [x] Support interactive removal of component defs
- [x] support renaming of morphs inside a component definition
- [x] fix `add()` being replaced with `without()`
- [x] enable collapse of unneeded sub specs on deep changes
- [x] Ensure spec props are in alphabetical order (except for `name`, `type` and `defaultViewModel`)
- [x] cleanup code
- [x] allow for different insertion points of submorphs instead of always appending to end
- [x] Ensure resolution by name paths instead of plain names.
- [x] Provide interface for enabling and disabling behavior in components
- [x] Improve button positioning via anchors.
- [x] Add proper opening of component defs via halo
- [x] fix proper ordered insert of submorphs on base component def
- [x] make reconciled code less verbose
- [x] fix ordering of uncollapsed sub specs
- [x] properly separate *live* instances and the editable component morph
- [x] fix incorrect applications of policies to instances where they are not the direct ancestor but just some distant relative.
- [x] properly reconcile font color properties inside text and attributes
- [x] Properly update the scene graph when there is a component replaced via code refresh.
- [x] Reset parts of derived component definitions that no longer "make sense" if one of the parents is being modified structurally.
- [x] add a sanitization mechanism for *broken* layout definitions once structure has changed
- [x] styled props by overridden master are not considered when overriding, comparing to original inherited value wich leads to incorrect reconciliation. e.g. => parent.fontSize = 12, self.master.fontSize = 16, self.fontSize = 12 -> no reconciliation 👎
- [x] support proper change propagation for inline policies

linusha commented:
When opening a component, the green button does not disappear for me:

![image](https://user-images.githubusercontent.com/14252419/194556635-060829d3-c5df-422a-b7f4-6f40e82e63ce.png)

Clicking into the text or scrolling fixes the issue.

linusha commented:
Opening a component and manipulating it with the sidebar (resulting in a new reconciled line being added) results in the buttons being placed like this:

![image](https://user-images.githubusercontent.com/14252419/194557075-d694e15b-8583-4cb0-9490-ddfff61c9c54.png)

linusha commented:
In general, editing text inside of the text morph that also contains the components buttons will hide the buttons. Is that intended?

My workflow right now is that I edit the ViewModel code while the component is opened in "editing mode" in the world.

linusha commented:
https://user-images.githubusercontent.com/14252419/194563726-1d7b2c65-72ba-41c3-94fb-0b4a6b2cc583.mp4

merryman commented:
> In general, editing text inside of the text morph that also contains the components buttons will hide the buttons. Is that intended?
> 
> My workflow right now is that I edit the ViewModel code while the component is opened in "editing mode" in the world.

We will implement a new solution that positions the buttons via anchors which are initially placed next to the component definitions. Assuming on initial load of the module we have a sound syntax. After that the positions of the edit buttons are controlled by the anchors.

linusha commented:
I just ran into a problem where the **none** of the border properties got reconciled (e.g., border color, width,...) but other properties (e.g. fill, rotation) worked as expected.

merryman commented:
> 

I cant seem to reproduce that.

merryman commented:
I am leaving this comment here as a memo for us regarding the discussion of *How are structural changes in parent components to be propagated among derived component definitions and their instances?*

**What is the problem?**
Right now the reconciliation works fairly well for real time adjustments to the _definitions of components in isolation_.
This means that we are able to reconcile changes to the source code and the runtime object corresponding to a component definition.
What we are lacking however is a policy of how these changes are to be propagated among derived components.
In the case of style property changes, this is very straight forward: We do not have to do anything since the application of policies takes care of the propagation among derived components. So only it really becomes tricky once we take into consideration structural changes that add or remove elements from a component.
This includes the source code reconciliation and in turn also the real time adjustments to the runtime objects.
For instance removing a morph in a component will do what to the adjustments in subsequent derived definitions? Are these to be deleted entirely? What happens to the current instances of these components? Are they also adjusted structurally? Applying a component style to a morph does not alter its structure. By which logic can we then propagate structural changes in the case of reconciliation? What happens when the morph we removed is inserted again (maybe at a different place.) Do derived morphs *remember* the adjustments that were just removed? If so on what basis? What happens in case of name collisions that arises in derived components but not in the parent we modify structurally? Is there a warning or is the operation prohibited?

**What are possible solutions?**
Looking at other tools such as Figma we observed some interesting ideas about how they reconcile structural changes among components that are derived from each other. A couple things stand out:
1. Figma places several restrictions on what can and can not be adjusted in the derived components. Locked properties include the structure of a derived component but also the position of its elements. Other things are fine to vary such as the fill or the border. Again other properties are *sort of in between* such as the adjustments one can do to the tiling layout (the axis can not be modified but things like spacing and padding can).
2. Figma in general does not allow structure to be modified in derived components. Varying structure in derived components is instead achieved via *nested components*. From the perspective of lively this would be as if `part()` calls inside a component definition became *slots* that are allowed to be replaced with entirely different `part()` calls once we work in a derived component definition. This simplifies the model greatly, however also comes at the cost of possibly more verbose component definition since there is less power in adjusting the structure of derived component definitions.
3. Although in the end a UI in Figma is always a document, where cycles do not exist, the runtime does a great deal of behind the scenes identity inference. For instance altering the structure in the parent of a family of components also affects the structure of the derived components. However Figma does in fact remember the removed substructures in the derived components and the fact that they were *associated* with the removed structure in the parent component. This can be regarded as a convenience that it attempts to maintain to a varying degree (it stops for instance once the removed structure is copied. At that point the associated elements are forgotten.)

**What is the likely best path forward?**

linusha commented:
# need to fix

https://user-images.githubusercontent.com/14252419/232024700-37f9fa02-df6c-4544-b99c-a9a9d5a55dbb.mp4

https://user-images.githubusercontent.com/14252419/232024730-b7eef3f6-ec59-4d3d-87e1-ec18b60dcab2.mp4

https://user-images.githubusercontent.com/14252419/232024819-11676abf-5d4e-4a73-8bf3-3ad5e7f896e9.mp4

https://user-images.githubusercontent.com/14252419/232024834-16e8ca2b-069e-4d91-8521-339ac76d9389.mp4

# nice to have (can easily be postponed)

- reset button is always clickable, even if there is nothing to reset and the cursor shows as forbidden

https://user-images.githubusercontent.com/14252419/232024892-0ff0356b-a237-475d-8e40-001dcc1b27e4.mp4

https://user-images.githubusercontent.com/14252419/232024906-d553a40f-db60-48c1-bead-2c660458a428.mp4

merryman commented on lively.ide/studio/controls/body.cp.js:
```diff
@@ -85,6 +85,8 @@ export class BodyControlModel extends PropertySectionModel {
    */
   ensureDynamicControls () {
     this.dynamicControls.forEach(m => m.remove());
+    // FIXME: adjusting the resize policies should automatically cause an override
+    this.view.layout = this.view.layout; // ensure the layout is overridden
```
This is really a layout problem. Currently in place modification of the layout does not really refresh reliably. We have to trigger the setter, since this is the only way it is going to ensure the layout is *applied*. We can go ahead and fix this, but I would argue that is a separate issue.

merryman commented on lively.halos/morph.js:
```diff
@@ -1,27 +1,35 @@
+import { getPropertiesNode } from 'lively.ide/components/helpers.js';
 
 const haloBlue = Color.rgb(23, 160, 251);
+const partAccent = Color.rgba(171, 71, 188, 1);
```
True. How about using `derivedAccent`? It basically marks anything that is part of a morph that was derived from a component.

linusha commented on lively.ide/studio/properties-panel.cp.js:
```diff
@@ -184,6 +184,12 @@ export class PropertiesPanelModel extends ViewModel {
   }
 
+  clearFocusIfRemoved () {
+    setTimeout(() => {
```
Can be removed

linusha commented on lively.morphic/layout.js:
```diff
@@ -640,6 +640,15 @@ export class TilingLayout extends Layout {
     } else super.onSubmorphResized(submorph, change);
   }
 
+  /**
+   * Invoked once a morph in the container has changed their bounds.
+   * @override
+   */
+  onSubmorphRemoved (submorph) {
+    this._resizePolicies.delete(submorph);
+    this.container.layout = this.copy();
+  }
+
```
Remove comment
copy for application and triggering of changed resize policies

linusha commented on lively.ide/js/browser/ui.cp.js:
```diff
@@ -8,6 +8,504 @@ import { EvalBackendButton } from '../eval-backend-ui.js';
+          return this.view?.owner;
+        }
+      },
+      behaviorEnabled: {
+        derived: true,
+        get () {
+          return this.editor?.owner.viewModel._initializeComponentsWithModels;
+        }
+      },
+      isLively: {
+        get () {
+          return !!this.instanceMorph;
+        }
+      },
```
We keep `isLively`

linusha commented on lively.ide/js/browser/ui.cp.js:
```diff
@@ -8,6 +8,504 @@ import { EvalBackendButton } from '../eval-backend-ui.js';
+    super.viewDidLoad();
+    this.updateResetButton();
+    once(this.componentDescriptor, 'makeDirty', this, 'updateResetButton');
+    connect(this.componentMorph, 'behaviorChanged', this, 'updateResetButton');
+  }
+
+  hasViewModels () {
+    let found = false;
+    withAllViewModelsDo(this.componentMorph, m => {
+      if (m.viewModel) found = true;
+    });
+    return found;
+  }
+
+  updateResetButton () {
```
Name gets changed

merryman commented:
One remaining issue occurs when multiple browser have the same component definition on display and we reconcile:
1. the changes are only reflected in one (seemingly random) browser instance
2. the file itself is not saved by the reconciliation. not even after the stop button has been pressed.

linusha commented on lively.ide/components/change-tracker.js:
```diff
@@ -0,0 +1,156 @@
+    if (!change.meta?.reconcileChanges) return true;
+    if (change.prop === 'name') return false;
+    if (change.prop === 'position' && (change.target === this.trackedComponent || this.isPositionedByLayout(change.target))) return true;
+    if (change.prop &&
+        change.prop !== 'textAndAttributes' &&
+        !change.target.styleProperties.includes(change.prop)) return true;
```
ignore non-style properties

linusha commented on lively.ide/components/change-tracker.js:
```diff
@@ -0,0 +1,156 @@
+    if (['addMorphAt', 'removeMorph'].includes(change.selector) &&
+        change.args.some(m => m.epiMorph)) return true;
+    if (!['addMorphAt', 'removeMorph'].includes(change.selector) && change.meta && change.meta.isLayoutAction) return true;
+    if (change.selector === 'addMorphAt' && change.target.textAndAttributes?.includes(change.args[0])) return true;
```
embedded morphs are handled by textAndAttributes, do not need to handle them again

linusha commented on lively.ide/components/editor.js:
```diff
@@ -0,0 +1,241 @@
+   * causes a name collision.
+   * Note that it does not matter wether or not the `addedMorph` is a entirely
+   * new morph or one that was reintroduced. At all times the renaming is applied
+   * to the `addedMorph`. This also means that derived adjustments will have to
+   * alter the name that they reference the `addedMorph` by if they are part
+   * of a reintroduction.
```
rausnehmen, drin was umbenennen und dann das alte wieder reinmachen. logisch, oder?

linusha commented on lively.ide/components/helpers.js:
```diff
@@ -0,0 +1,398 @@
+         ]
+       ]
+  `);
+  // always pick the one closest
+  // if (morphDefs.length > 1) throw new Error('ambigous name reference: ' + name);
+  return morphDefs[0];
+}
```
remove if statement

linusha commented:
Please expand https://github.com/LivelyKernel/lively.next/wiki/Components as well

linusha commented on lively.ide/components/change-tracker.js:
```diff
@@ -0,0 +1,156 @@
+   * @returns { boolean }
+   */
+  adjournChange (change) {
+    const isReplaceChange = change.selector === 'replace';
+    if (!isReplaceChange) return false;
+    const insertsMorph = change.args[1].find(m => m?.isMorph);
+    const removesMorph = change.undo.args[1].find(m => m?.isMorph);
+    return !insertsMorph && !removesMorph;
```
@merryman needs to be enabled again

linusha reviewed (changes requested):
1. The 'lively' halo and button do not display the same behavior.
2. Many tests in morphic/component are failing
